File: Solutions/Cpp/valid_palindrome.cpp
/*
    [EASY]
    125. Valid Palindrome

    Concepts:
    - string
    - two pointers

    Stats:
        Runtime | 3 ms      [Beats 32.21%]
        Memory  | 10.29 MB  [Beats 27.99%]
*/

class Solution {
public:
    bool isPalindrome(string s) {
        std::vector<char> vec = {};
        for (char c : s) {
            if (std::isalnum(c))
                vec.push_back(tolower(c));
        }

        int n = vec.size();
        for (int i = 0; i < (n + 1) / 2; i++) {
            if (vec[i] != vec[n - i - 1])
                return false;
        }
        return true;
    }
};

File: Solutions/Python/valid_palindrome.py
"""
    [EASY]
    125. Valid Palindrome

    Concepts:
    - string
    - two pointers

    Stats:
        Runtime | 4 ms      [Beats 90.75%]
        Memory  | 18.40 MB  [Beats 41.66%]
"""

class Solution:
    def isPalindrome(self, s: str) -> bool:
        s = "".join(list(filter(str.isalnum, s))).lower()
        n = len(s)

        for i in range((n + 1) // 2):
            if s[i] != s[-(i+1)]:
                return False
        return True